<?xml version="1.0" encoding="UTF-8"?>
<root BTCPP_format="3"
      main_tree_to_execute="MainTree">
  <BehaviorTree ID="MainTree">
    <Fallback>
      <TimingDecorator name="Total">
        <Sequence>
          <Sequence>
            <SetBlackboard value="-0.35,0.0,0.0"
                           output_key="pre_grab_box_offset"/>
            <SetBlackboard value="-0.3,0.0,0.0"
                           output_key="pre_grab_box_offset_after_sgl_step"/>
            <SetBlackboard value="0.35,0.35,0.3"
                           output_key="box_size"/>
            <SetBlackboard value="-0.50,0,0"
                           output_key="pre_grab_box_offset2"/>
            <SetBlackboard value="0,0,0"
                           output_key="force_threshold"/>
          </Sequence>
          <CheckStatusOK/>
          <ArmMoveToReadyPose action_name="ready_joints"
                              move_speed="1.0"/>
          <SequenceStar>
            <SequenceStar>
              <TimingDecorator name="MoveToGraspCMD">
                <ReactiveSequence>
                  <RetryUntilSuccessful num_attempts="3">
                    <BoxTagOK tag_id="1"
                              box_pos="{box_pose}"
                              box_quat="{box_quat}"
                              time_check="false"/>
                  </RetryUntilSuccessful>
                  <ComputeTargetPose box_pos="{box_pose}"
                                     box_quat="{box_quat}"
                                     target_box_offset="{pre_grab_box_offset}"
                                     target_pose="{target_pose}"/>
                  <CmdPoseWorldMoveToDestination target_pose="{target_pose}"/>
                </ReactiveSequence>
              </TimingDecorator>
              <TimingDecorator name="Grasp">
                <Sequence>
                  <Sequence>
                    <RetryUntilSuccessful num_attempts="10">
                      <BoxTagOK tag_id="1"
                                box_pos="{box_pose}"
                                box_quat="{box_quat}"
                                time_check="true"/>
                    </RetryUntilSuccessful>
                    <ComputeTargetPose box_pos="{box_pose}"
                                       box_quat="{box_quat}"
                                       target_box_offset="{pre_grab_box_offset_after_sgl_step}"
                                       target_pose="{target_pose_pre_grasp}"/>
                  </Sequence>
                  <RetryUntilSuccessful num_attempts="3">
                    <GraspBox box_pos="{box_pose}"
                              box_quat="{box_quat}"
                              box_size="{box_size}"
                              box_offset="-0.05,0.0,0.16"
                              grasp_type="0"/>
                  </RetryUntilSuccessful>
                  <MoveToDestination planed_trajectory="0.0, 0.0, 0.0, 0.0"
                                     move_type="2"/>
                  <SetBlackboard value="0,0,0"
                                 output_key="force_threshold"/>
                </Sequence>
              </TimingDecorator>
              <TimingDecorator name="GraspBcakCMD">
                <Sequence>
                  <MakePlan target_pose="{target_pose_pre_put}"
                            planed_trajectory="{planed_trajectory_pre_put}"/>
                  <ComputeTargetPose box_pos="{box_pose}"
                                     box_quat="{box_quat}"
                                     target_box_offset="-1.0,0,0"
                                     target_pose="{target_pose_pre_put}"/>

                  <!-- <CmdPoseWorldMoveToDestination target_pose="{target_pose_pre_put}"/> -->

                </Sequence>
              </TimingDecorator>
            </SequenceStar>
            <SequenceStar>
              <TimingDecorator name="MoveToPutSingle(Turn)">
                <Sequence>
                  <RetryUntilSuccessful num_attempts="3">
                    <BoxTagOK tag_id="2"
                              box_pos="{box_pose2}"
                              box_quat="{box_quat2}"
                              time_check="false"/>
                  </RetryUntilSuccessful>
                </Sequence>
              </TimingDecorator>
              <TimingDecorator name="MoveToPutCMD">
                <ReactiveSequence>
                  <RetryUntilSuccessful num_attempts="3">
                    <BoxTagOK tag_id="2"
                              box_pos="{box_pose2}"
                              box_quat="{box_quat2}"
                              time_check="false"/>
                  </RetryUntilSuccessful>
                  

                  <ComputeTargetPose box_pos="{box_pose2}"
                                     box_quat="{box_quat2}"
                                     target_box_offset="{pre_grab_box_offset2}"
                                     target_pose="{target_pose_pre_put}"/>
                  <CmdPoseWorldMoveToDestination target_pose="{target_pose_pre_put}"/>
                </ReactiveSequence>
              </TimingDecorator>
              <Sequence>
                <TimingDecorator name="Put">
                  <Sequence>
                    <RetryUntilSuccessful num_attempts="3">
                      <BoxTagOK tag_id="2"
                                box_pos="{box_pose2}"
                                box_quat="{box_quat2}"
                                time_check="false"/>
                    </RetryUntilSuccessful>
                    <RetryUntilSuccessful num_attempts="5">
                      <GraspBox box_pos="{box_pose2}"
                                box_quat="{box_quat2}"
                                box_size="{box_size}"
                                box_offset="-0.13,0.0,-0.55"
                                grasp_type="1"/>
                    </RetryUntilSuccessful>
                  </Sequence>
                </TimingDecorator>
                <MoveToDestination planed_trajectory="0.0, 0.0, 0.0, 0.0"
                                   move_type="2"/>
                <SetBlackboard value="0,0,0"
                               output_key="force_threshold"/>
                <Parallel failure_threshold="1"
                          success_threshold="2">
                  <TimingDecorator name="PutBackCMD">
                    <Sequence>
                      <ComputeTargetPose box_pos="{box_pose2}"
                                         box_quat="{box_quat2}"
                                         target_box_offset="-1.0,0,0"
                                         target_pose="{target_pose_pre_put}"/>
                      <MakePlan target_pose="{target_pose_pre_put}"
                                planed_trajectory="{planed_trajectory2}"/>
                      <MoveToDestination planed_trajectory="{planed_trajectory2}"
                                         move_type="1"/>
                    </Sequence>
                  </TimingDecorator>
                  <TimingDecorator name="ArmHome">
                    <Delay delay_msec="2000">
                      <ArmMoveToReadyPose action_name="home_joints"
                                          move_speed="1.0"/>
                    </Delay>
                  </TimingDecorator>
                </Parallel>
              </Sequence>
            </SequenceStar>
          </SequenceStar>
        </Sequence>
      </TimingDecorator>
    </Fallback>
  </BehaviorTree>

  <!-- Description of Node Models (used by Groot) -->
  <TreeNodesModel>
    <Action ID="ArmMoveToReadyPose"
            editable="true">
      <input_port name="action_name"/>
      <input_port name="move_speed"/>
    </Action>
    <Condition ID="BoxTagOK"
               editable="true">
      <input_port name="tag_id"/>
      <output_port name="box_pos"/>
      <output_port name="box_quat"/>
      <input_port name="time_check"
                  default="false"/>
    </Condition>
    <Condition ID="CheckStatusOK"
               editable="true"/>
    <Action ID="CmdPoseWorldMoveToDestination">
      <input_port name="target_pose"/>
    </Action>
    <Condition ID="ComputeTargetPose"
               editable="true">
      <input_port name="box_pos"/>
      <input_port name="box_quat"/>
      <input_port name="target_box_offset"/>
      <output_port name="target_pose"/>
    </Condition>
    <Action ID="GraspBox"
            editable="true">
      <input_port name="box_pos"/>
      <input_port name="box_quat">xyzw</input_port>
      <input_port name="box_size"/>
      <input_port name="box_offset"/>
      <input_port name="grasp_type">0:grasp, 1:put</input_port>
    </Action>
    <Action ID="MakePlan"
            editable="true">
      <input_port name="target_pose"/>
      <output_port name="planed_trajectory"/>
    </Action>
    <Action ID="MoveToDestination"
            editable="true">
      <input_port name="planed_trajectory"/>
      <input_port name="move_type"
                  default="0">int</input_port>
    </Action>
    <Decorator ID="TimingDecorator"
               editable="true"/>
  </TreeNodesModel>

</root>
